File: src/main/org/nlogo/awt/Clipboard.java
package org.nlogo.awt;

import java.awt.Toolkit;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.awt.datatransfer.UnsupportedFlavorException;

public final strictfp class Clipboard {

  // this class is not instantiable
  private Clipboard() { throw new IllegalStateException(); }

  /// clipboard

  public static String getClipboardAsString(Object requester) {
    Toolkit kit = Toolkit.getDefaultToolkit();
    java.awt.datatransfer.Clipboard clipboard = kit.getSystemClipboard();
    Transferable transferable = clipboard.getContents(requester);
    if (transferable == null) {
      return null;
    }
    try {
      return (String) transferable.getTransferData(DataFlavor.stringFlavor);
    } catch (java.io.IOException ex) {
      // ignore exception
      return null;
    } catch (UnsupportedFlavorException ex) {
      // ignore exception
      return null;
    }
  }

}

File: src/main/org/nlogo/awt/Clipboard.scala
package org.nlogo.awt

import java.awt.Toolkit
import java.awt.datatransfer.{ DataFlavor, Transferable, UnsupportedFlavorException }

object Clipboard {
  def getClipboardAsString(requester: AnyRef): String = {
    val kit = Toolkit.getDefaultToolkit
    val clipboard = kit.getSystemClipboard
    val transferable = clipboard.getContents(requester)
    def getString(t: Transferable) =
      t.getTransferData(DataFlavor.stringFlavor).asInstanceOf[String]
    try Option(transferable).map(getString).orNull
    catch {
      case _: java.io.IOException | _: UnsupportedFlavorException =>
        null
    }
  }
}
